### PR TITLE
Add opentracing support

### PR DESCRIPTION
**Pull request by gnarula**
Adds support for tracing GRPC calls using Jaeger. The implementation allows instantiating one tracer per node, identified by its address.  A process may therefore have multiple tracers, specially in non production settings where multiple nodes run in the same process.

### CODE REVIEW

**Comment by gnarula** [comment]
A quick way to try this out is to run jaeger locally using:

```bash
$ docker run --name jaeger \
  -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
  -p 5775:5775/udp \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 16686:16686 \
  -p 14268:14268 \
  -p 14250:14250 \
  -p 9411:9411 \
 -it  jaegertracing/all-in-one:1.20 --log-level=debug
```

and then run the `ExampleRPC_Stream` test in `/mino/minogrpc`:

```bash
$ export JAEGER_ENDPOINT=http://localhost:14268/api/traces
$ export JAEGER_SAMPLER_TYPE=const
$ export JAEGER_SAMPLER_PARAM=1
$ go test -v -run ExampleRPC_Stream ./...
```

If you do not wish to run the test, you may import a trace that I ran by importing the following trace JSON https://paste.ubuntu.com/p/4fzk9zkGDW/

P.S. Jaeger's UI is accessible at http://localhost:16686/

File: mino/minogrpc/example_test.go
Comment (pierluca) [inline mino/minogrpc/example_test.go:60]: I would maintain the previous example for the sake of clarity, as it's simpler and much needed to explain the basic Stream setup.

I understand the new code is very nice to test the system, but I would possibly create a separate Example, possibly mentioning its purpose, e.g. `ExampleRPC_OpenTracingDemo` and adding some comments as to how to see all its opentracing goodness

File: mino/minogrpc/server.go
Comment (pierluca) [inline mino/minogrpc/server.go:675]: This function is getting humongous. 
Consider splitting lines 645-667 into a separate function,

File: mino/minogrpc/server.go
Comment (pierluca) [inline mino/minogrpc/server.go:696]: does this function need to be declared inline? I'm not sure I see why.
Could we deport it outside this function, again for readability?

File: mino/minogrpc/rpc.go
Comment (pierluca) [inline mino/minogrpc/rpc.go:155]: this "undefined" looks like a magic value, which could actually match a protocol's name.

I would suggest creating a constant with an unlikely protocol name, such as `__UNDEFINED_PROTOCOL__` or similar

File: mino/minogrpc/tracing/tracers.go
Comment (pierluca) [inline mino/minogrpc/tracing/tracers.go:20]: naming: this doesn't represent a `tracer` but a (e.g.) `tracerCatalog` right?

File: mino/minogrpc/tracing/tracers.go
Comment (pierluca) [inline mino/minogrpc/tracing/tracers.go:30]: naming: ...and this would thus just be our `catalog`, right?

File: mino/minogrpc/tracing/tracers.go
Comment (pierluca) [inline mino/minogrpc/tracing/tracers.go:25]: naming: this is not a structure that can close or closes tracers, it's a "closable" tracer, right?
Why not just call it `closableTracer` or `tracerWrapper` ?

File: mino/minogrpc/tracing/tracers.go
Comment (pierluca) [inline mino/minogrpc/tracing/tracers.go:1]: I love the elegance of this simple file.

However, I had to re-read it many times and found the naming very confusing.
I've added a ton of comments, feel free to ignore them altogether if you feel I'm wrong :)

File: mino/minogrpc/tracing/tracers.go
Comment (pierluca) [inline mino/minogrpc/tracing/tracers.go:21]: `tracerByAddr` could be more explicit.

**Review by pierluca — COMMENTED** [review]
A few remarks for maintainability, but I look forward to using this feature :)

File: mino/minogrpc/mod.go
Comment (nkcr) [inline mino/minogrpc/mod.go:271]: error wrapping

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:13]: exported variable should have a comment

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:18]: exported const should have a comment

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:16]: is it necessary to define and use that type ?

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:36]: you could directly define tracer, which removes the need of using init:
```
var tracerInstance = tracer{
	addrTracer: make(map[string]tracerCloser),
}
```

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:52]: please, systematically wrap errors

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:64]: I would add a blank line

File: mino/minogrpc/tracing/tracers.go
Comment (nkcr) [inline mino/minogrpc/tracing/tracers.go:60]: I would add a blank line

**Review by nkcr — COMMENTED** [review]
I tried the tracing and it looks nice 😀! Though I must admit I would need a crash course to understand how to efficiently use it.

It would be better to move the opentracing implementation in `internal/` and remove its coupling to minogrpc. It would also help other module use it, without having to be aware of a specific mino implementation like it is the case right now.

File: mino/minogrpc/tracing/tracers.go
Comment (gnarula) [reply mino/minogrpc/tracing/tracers.go:20]: Exactly, `catalog` was the word I was looking for!

File: mino/minogrpc/tracing/tracers.go
Comment (gnarula) [reply mino/minogrpc/tracing/tracers.go:16]: IIRC it had to do with the linter complaining about possible key clashes in context.Value

**Review by pierluca — COMMENTED** [review]
This looks much better. Great !

**Comment by nkcr** [comment]
Cool, looks good :). Please let me do a second quick pass before merging.

File: core/ordering/cosipbft/mod.go
Comment (nkcr) [inline core/ordering/cosipbft/mod.go:550]: Is it necessary ? `s.synx.Sync` will do the same.

File: core/ordering/cosipbft/blocksync/default.go
Comment (nkcr) [inline core/ordering/cosipbft/blocksync/default.go:28]: Do you really need to have it publicly declared ? The only place is in cosipbft/mod.go, which seems superfluous.

File: cosi/threshold/mod.go
Comment (nkcr) [inline cosi/threshold/mod.go:25]: Apparently it doesn't need to be public.

File: dkg/pedersen/mod.go
Comment (nkcr) [inline dkg/pedersen/mod.go:31]: Same, do we need to have it public ?

File: internal/tracing/tracers.go
Comment (nkcr) [inline internal/tracing/tracers.go:23]: wondering if just having "UNDEFINED" wouldn't do the job ?

File: internal/tracing/tracers.go
Comment (nkcr) [inline internal/tracing/tracers.go:28]: A detail, but we tend to always place sync.Mutex as the first attribute.

File: internal/tracing/tracers.go
Comment (nkcr) [inline internal/tracing/tracers.go:54]: Again, a detail, but for consistency we normally always use "failed to..." or sometimes "couldn't ...".
Same bellow.

File: internal/tracing/tracers.go
Comment (nkcr) [inline internal/tracing/tracers.go:76]: error wrapping

File: mino/minogrpc/mod.go
Comment (nkcr) [inline mino/minogrpc/mod.go:176]: Error wrapping. Also we should make sure this new path is covered in testing.

File: mino/minogrpc/mod.go
Comment (nkcr) [inline mino/minogrpc/mod.go:254]: for consistency, use rather `failed to close tracers: %v`.

File: mino/minogrpc/mod.go
Comment (nkcr) [inline mino/minogrpc/mod.go:339]: I would comment what this function is used for.

File: mino/minogrpc/mod.go
Comment (nkcr) [inline mino/minogrpc/mod.go:344]: I would add a blank line there.

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:43]: Do we still need to have it public ?

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:647]: "failed to..."

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:659]: error wrapping. Also be sure the new path is covered in testing.

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:780]: blank line

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:774]: I would add a quick comment. Also a function's name should describe an action with a verb. Maybe `decorateClientTrace` ?

File: mino/minogrpc/server.go
Comment (nkcr) [inline mino/minogrpc/server.go:685]: Wondering if `getCredential` wouldn't be better.

File: mino/minogrpc/mod.go
Comment (nkcr) [reply mino/minogrpc/mod.go:339]: Also a function's name should preferable describe an action, with a verb. Maybe `decorateServerTrace` ?

**Review by nkcr — COMMENTED** [review]
This is mostly about conventions. We've put a lot of effort with Gaylor to keep consistency and conventions. Most of them are very opinionated (and could arguably be different!). But it proved to be very beneficial so far. They can be found there https://dedis.github.io/dela/#/guideline, and I will try to update them along the way.

File: core/ordering/cosipbft/mod.go
Comment (gnarula) [reply core/ordering/cosipbft/mod.go:550]: Nice catch! Removing this.

File: core/ordering/cosipbft/blocksync/default.go
Comment (gnarula) [reply core/ordering/cosipbft/blocksync/default.go:28]: Made it public because it was used in s.sync.Sync which now seems redundant. Gonna make these and the ones below private

File: internal/tracing/tracers.go
Comment (gnarula) [reply internal/tracing/tracers.go:23]: I reckon @pierluca's review remarked `undefined` to be ambiguous. I think `__UNDEFINED_PROTOCOL__` is fine - it stands out and might make it easier to notice because we'd want to name all protocols in the traces.

File: mino/minogrpc/mod_test.go
Comment (nkcr) [inline mino/minogrpc/mod_test.go:96]: you can use `fake.Error("failed to get tracer for addr 127.0.0.1:3333")`, which does the `fmt.Sptintf` for you.

**Review by nkcr — APPROVED** [review]
Looks very nice!

**Comment by nkcr** [comment]
can be merged once it's updated with master

**Comment by gnarula** [comment]
@nkcr rebased with master. Ready for merge now :)

**Comment by nkcr** [comment]
So cool 😀

**Pull request merged**
